libsupport/include/katana/Result.h

```diff
@@ -260,6 +260,40 @@ ResultErrno() {
   return std::error_code(errno, std::system_category());
 }
 
+template <class>
+struct is_result : std::false_type {};
+
+template <class T>
+struct is_result<Result<T>> : std::true_type {};
+
+template <class T>
+std::enable_if_t<!std::is_same<T, void>::value, T&&>
+extract_value(Result<T>&& result) {
+  return std::move(result.value());
+}
+
+inline int
+extract_value(Result<void>&&) {
+  return 0;
+}
+
+#define KATANA_CHECK_NAME(x, y) x##y
+
+#define KATANA_CHECK_IMPL(result_name, expression)                             \
+  ({                                                                           \
+    auto result_name = (expression);                                           \
+    static_assert(                                                             \
+        ::katana::is_result<decltype(result_name)>::value,                     \
+        "KATANA_CHECK requires a katana::Result");                             \
+    if (!result_name) {                                                        \
+      return result_name.error().WithContext("here");                          \
+    }                                                                          \
+    std::move(::katana::extract_value(std::move(result_name)));                \
+  })
```

Comment:
Neat! I thought `({....})`this was still a GCC only extension, it seems like it passes all of the cases that we care about.

Reply:
not standard

Reply:
It's like `__attribute__`. It's not standard, but is very common. I don't know if niche compilers like ICC and MSVC support it. But the main compilers out there do. (Yes, I called MSVC a niche compiler, fight me. ;-) )

Reply:
Last I saw ICC supports it with caveats, the main one being objects with destructors cannot be initialized inside. We'll cross that bridge when we come to it

Reply:
What's the issue with destructors? I could see how this could break scope extension. Is there anything other than lacking scope extension?

Reply:
Is `do {  ..... } while (0)` a solution? I frequently see this idiom when building multiple line macros.

Reply:
Oh, I see the problem here: it wants to return the last line.

Reply:
And the `do {} while()` idiom is still a statement. This needs to be an expression.

Reply:
ICC has several caveats with their statement expressions described [here](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compatibility-and-portability/gcc-compatibility-and-interoperability.html). Since that page concerns their gcc compatibility layer, I assume those caveats spring from the fact that it's probably an ad-hoc  construction in their AST.

And yeah, `do/while` wouldn't be usable here as an outer wrapper. What the statement expression is giving us is that the last expression inside is treated as the return value for the whole block. So it's already self-contained, and wouldn't be able to work inside a `do/while`

Reply:
IBM XLC also supports statement expressions, so MSVC is more niche than that